packages/iov-bns-governance/src/utils.ts

```diff
@@ -0,0 +1,27 @@
+export function groupByCallback<T, K>(
+  values: readonly T[],
+  callback: (value: T) => K,
+  // tslint:disable-next-line:readonly-array readonly-keyword
+): { readonly key: K; values: T[] }[] {
+  return values.reduce(
+    (grouped, value) => {
+      const key = callback(value);
+      const existing = grouped.find(group => group.key === key);
+      if (existing) {
+        existing.values.push(value);
+      } else {
+        grouped.push({
+          key: key,
+          values: [value],
+        });
+      }
+      return grouped;
+    },
+    // tslint:disable-next-line:readonly-array readonly-keyword
+    [] as { readonly key: K; values: T[] }[],
+  );
+}
+
+export function maxWithComparatorCallback<T>(values: readonly T[], callback: (a: T, b: T) => T): T {
+  return values.slice(1).reduce(callback, values[0]);
```

Comment:
Pretty close to what I has in mind. Ideas:
1. A typical comparator function (across many different ecosystems) returns a negative value, 0, a positive value (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#Description), allowing implementations like `(rule1, rule2) => rule1.version - rule2.version`
2. `maxWithComparatorCallback` can be implemented using array.sort and using the last element
3. There should be some kind of defined behavior for `values` being empts. A hard throw is fine I guess.

Reply:
Re 2, I ended up with a solution which was pretty simple even without using sort. Probably we won't have enough proposals anytime soon to make the complexity difference significant, but just in case we'll be ready.

Reply:
Sometimes it is about the beauty of code :)